src/docker/shared.rs

```diff
@@ -294,7 +294,7 @@ impl Directories {
         let xargo = file::canonicalize(&xargo)?;
         let nix_store = match nix_store {
             Some(store) => Some(file::canonicalize(&store)?),
-            None => None,
+            None => Some(PathBuf::from(r"/nix/store")),
         };
```

Comment:
First, a minor nit: this doesn't need to be a raw string. Also, I think we should only provide a default if the host OS is Linux. Something like:

```rust
let nix_store = match nix_store {
    Some(store) => Some(file::canonicalize(&store)?),
    None if cfg!(target_os = "linux") => Some(PathBuf::from("/nix/store")),
    None => None,
};
```

Reply:
Also, the following needs to occur after providing the default if we do modify this:

```
        if let Some(ref nix_store) = nix_store {
            file::create_dir_all(nix_store)?;
        }
```

This might be an issue, since if it's not as root it will fail, and then we can't mount the nix store. IE, we can only bind-mount it if we have permissions to write there as the current user.

Reply:
I think it makes sense to check for aliveness and only specify a -v mount if the folder exists

Reply:
e.g, I don't want a `/nix/store` folder to be created on every system that uses cross.

Reply:
Ok then probably something like this?

```rust
let nix_store = env::var_os("NIX_STORE_DIR")
    .or_else(|_| env::var_os("NIX_STORE"))
    .map(PathBuf::from);

let nix_store = match nix_store {
    Some(store) => {
        let path = file::canonicalize(&store)?;
        fs::create_dir_all(&path)?;
        Some(path)
    }
    None if cfg!(target_os = "linux") && fs::exists("/nix/store") => Some(PathBuf::from("/nix/store")),
    None => None,
};
```

This way we now support the `NIX_STORE_DIR` override, fallback to `NIX_STORE`, and only provide the default if we're on Linux and the directory exists (so we don't make a filesystem call unless required).

Reply:
I imagine it'd work.  I wonder why we are calling the `fs::create_dir_all` and if it'd be needed in both cases to allow mounting the volume.

Reply:
> I imagine it'd work. I wonder why we are calling the `fs::create_dir_all` and if it'd be needed in both cases to allow mounting the volume.

Because we want it to fail if it doesn't exist and we don't have write permissions, since Docker will create the directory as root if it doesn't exist (or any container engine running as root). We could also just error if `NIX_STORE` or `NIX_STORE_DIR` is provided but the directory doesn't exist (and not error if an implicit `/nix/store` doesn't exist).

Reply:
I tried this, but I am new to rust so I might be missing something. Here is the error I got:
```
error[E0425]: cannot find function `exists` in module `fs`
   --> src/docker/shared.rs:304:54
    |
304 |             None if cfg!(target_os = "linux") && fs::exists("/nix/store") => Some(PathBuf::from("/nix/store")),
    |                                                      ^^^^^^ not found in `fs`

error[E0593]: closure is expected to take 0 arguments, but it takes 1 argument
   --> src/docker/shared.rs:275:14
    |
275 |             .or_else(|_| env::var_os("NIX_STORE"))
    |              ^^^^^^^ --- takes 1 argument
    |              |
    |              expected closure that takes 0 arguments
```

Reply:
You can try:

```rust
let default_nix_store = PathBuf::from("/nix/store");

let nix_store = env::var_os("NIX_STORE_DIR")
    .or_else(|| env::var_os("NIX_STORE"))
    .map(PathBuf::from);

let nix_store = match nix_store {
    Some(store) => {
        let path = file::canonicalize(&store)?;
        fs::create_dir_all(&path)?;
        Some(path)
    }
    None if cfg!(target_os = "linux") && default_nix_store.exists() => Some(default_nix_store),
    None => None,
};
```

Reply:
That one works!

Reply:
@IllustratedMan-code This is what I was thinking:

```diff
diff --git a/src/docker/shared.rs b/src/docker/shared.rs
index 9fb7675..09632ce 100644
--- a/src/docker/shared.rs
+++ b/src/docker/shared.rs
@@ -270,7 +270,10 @@ impl Directories {
         let cargo = home::cargo_home()?;
         let xargo =
             env::var_os("XARGO_HOME").map_or_else(|| home_dir.join(".xargo"), PathBuf::from);
-        let nix_store = env::var_os("NIX_STORE").map(PathBuf::from);
+        // NIX_STORE_DIR is an override of NIX_STORE, which is the path in derivations.
+        let nix_store = env::var_os("NIX_STORE_DIR")
+            .or_else(|| env::var_os("NIX_STORE"))
+            .map(PathBuf::from);
         let target = &metadata.target_directory;
 
         // create the directories we are going to mount before we mount them,
@@ -279,9 +282,6 @@ impl Directories {
         // if it has other issues (such as permission errors).
         file::create_dir_all(&cargo)?;
         file::create_dir_all(&xargo)?;
-        if let Some(ref nix_store) = nix_store {
-            file::create_dir_all(nix_store)?;
-        }
         create_target_dir(target)?;
 
         // get our mount paths prior to canonicalizing them
@@ -294,6 +294,20 @@ impl Directories {
         let xargo = file::canonicalize(&xargo)?;
         let nix_store = match nix_store {
             Some(store) => Some(file::canonicalize(&store)?),
+            None => Some(PathBuf::from("/nix/store")),
+        };
+
+        let default_nix_store = PathBuf::from("/nix/store");
+
+        let nix_store = match nix_store {
+            Some(store) if store.exists() => {
+                let path = file::canonicalize(&store)?;
+                Some(path)
+            }
+            Some(store) => {
+                eyre::bail!("unable to find provided nix-store directory {store:?}");
+            }
+            None if cfg!(target_os = "linux") && default_nix_store.exists() => Some(default_nix_store),
             None => None,
         };
 
```

Reply:
Looks good to me! I'll go ahead and commit a changelog file as well as the changes you suggested.